packages/core/src/browser/frontend-connection-status.ts

```diff
@@ -0,0 +1,356 @@
+            updated.shift();
+        }
+        return updated;
+    }
+
+}
+
+@injectable()
+export class ConnectionStatusStatusBarContribution extends DefaultFrontendApplicationContribution {
+
+    private idle: boolean;
+
+    constructor(
+        @inject(ConnectionStatusService) protected readonly connectionStatusService: ConnectionStatusService,
+        @inject(FrontendIdleService) protected idleService: FrontendIdleService,
+        @inject(StatusBar) protected statusBar: StatusBar
+    ) {
+        super();
+        this.connectionStatusService.onStatusChange(event => this.onStatusChange(event));
+        this.idle = this.idleService.idle;
+        this.idleService.onStatusChange(idle => {
+            if (this.idle !== idle) {
+                this.idle = idle;
+                if (this.idle) {
+                    this.statusBar.removeElement('connection-status');
+                    this.statusBar.setElement('connection-status', {
+                        alignment: StatusBarAlignment.RIGHT,
+                        text: '$(bed)',
+                        priority: 0,
+                        tooltip: 'Idle'
+                    });
+                } else {
+                    this.onStatusChange(this.connectionStatusService.currentState);
+                }
+            }
+        });
+    }
+
+    protected onStatusChange(event: ConnectionStatus) {
+        if (!this.idle) {
+            this.statusBar.removeElement('connection-status');
+            const text = `$(${this.getStatusIcon(event.health)})`;
+            const tooltip = event.health ? `Online [Connection health: ${event.health}%]` : 'Offline';
+            this.statusBar.setElement('connection-status', {
+                alignment: StatusBarAlignment.RIGHT,
+                text,
+                priority: 0,
+                tooltip
+            });
+        }
+    }
+
+    private getStatusIcon(health: number | undefined) {
+        if (health === undefined || health === 0) {
+            return 'exclamation-circle';
+        }
+        if (health < 25) {
+            return 'frown-o';
+        }
+        if (health < 50) {
+            return 'meh-o';
+        }
+        return 'smile-o';
+    }
+
+}
+
+@injectable()
+export class ApplicationConnectionStatusContribution extends DefaultFrontendApplicationContribution {
+
+    private dialog: ConnectionStatusDialog | undefined;
+    private state = ConnectionState.ONLINE;
+
+    constructor(
+        @inject(ConnectionStatusService) protected readonly connectionStatusService: ConnectionStatusService,
+        @inject(MessageService) protected readonly messageService: MessageService,
+        @inject(ILogger) protected readonly logger: ILogger
+    ) {
+        super();
+        this.connectionStatusService.onStatusChange(event => this.onStatusChange(event));
+    }
+
+    protected onStatusChange(event: ConnectionStatus): void {
+        if (this.state !== event.state) {
```

Comment:
It can happen that on `connectionStatusService.onStatusChange` is called with the same value several times? I think it should be called only when a value changed that clients don't do prev state caching and comparison like here.

Reply:
As you wish.

Reply:
This has to be there. It will not change.

Reply:
Yeah, otherwise, I cannot adjust the log message. I cannot make difference if the state was idle or offline previously after receiving the online state. Furthermore, the connection health which is a percentage between 0 and 100 can change while the state is still online. I hope it explains.

Reply:
Or I can provide another hook so that we would have both the `onConnectionStatusChange` and an `onConnectionStatusStateChange`.